### PR TITLE
Fix AOTAutograd cache save for call-free lazy backwards

### LINKED ISSUE #141790: AOTAutograd cache always misses even on simple program
### 🐛 Describe the bug

```
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (a1348a73)]$ TORCHINDUCTOR_AUTOGRAD_CACHE=1 TORCH_LOGS=torch._functorch._aot_autograd.autograd_cache python aaa.py 
I1128 17:54:16.961000 1579551 torch/_functorch/_aot_autograd/autograd_cache.py:665] [0/0] AOTAutograd cache miss for key a3fjisonhscsotmkqcifvx22whts72lyron5lrhskrbowx3hfbav
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (a1348a73)]$ TORCHINDUCTOR_AUTOGRAD_CACHE=1 TORCH_LOGS=torch._functorch._aot_autograd.autograd_cache python aaa.py 
I1128 17:54:26.680000 1582644 torch/_functorch/_aot_autograd/autograd_cache.py:665] [0/0] AOTAutograd cache miss for key a3fjisonhscsotmkqcifvx22whts72lyron5lrhskrbowx3hfbav
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (a1348a73)]$ vim aaa.py
(/home/ezyang/local/b/pytorch-env) [ezyang@devgpu005.nha1 ~/local/b/pytorch (a1348a73)]$ cat aaa.py
import torch
import torch._dynamo.config



# Artificially generate lots of small kernels
torch._inductor.config.realize_reads_threshold = 1
torch._inductor.config.realize_opcount_threshold = 1
torch._inductor.config.max_fusion_size = 1


N = 2


@torch.compile(fullgraph=True)
def f(a, b):
    for i in range(N):
        a = a + b * i
    return a


f(torch.randn(2, device="cuda", requires_grad=True), torch.randn(2, device="cuda")).sum().backward()
```

### Versions

main

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186053

AOTAutograd local cache entries store FXGraphCache keys for the compiled forward and backward artifacts. The issue reproducer has a lazy backward graph that contains no call_function/call_module nodes and just returns the incoming tangent. Inductor handled that backward through its no-op fast path, returning a boxed Python function before FXGraphCache ran. That function had no _fx_graph_cache_key, so AOTAutograd's save hook declined to write an entry and the same cache key missed on every process run.

Fix this by threading an explicit per-compile signal from AOTAutograd cache miss handling into Inductor. When a valid AOTAutograd miss is preparing a saveable entry, call-free forward/backward compiles skip the no-op shortcut and go through FXGraphCache so the resulting artifact carries the key AOTAutograd needs. Cache bypasses and runs without FXGraphCache keep the existing fast path. The control flag is excluded from FXGraphCache hashing because it does not affect generated code for normal graphs.

This keeps the existing full forward/backward cache-entry contract rather than introducing partial AOTAutograd entries or broadly eager-compiling lazy backwards.

Fixes #141790
Generated by my agent

Benchmark Results:
On the issue reproducer, measured `torch.compile(fn, backend="inductor", fullgraph=True)(a, b).sum().backward()` for 10 cache-miss iterations on CPU with fx_graph_cache=True, fx_graph_remote_cache=False, compile_threads=1, enable_autograd_cache=True, enable_remote_autograd_cache=False.

Before:
- times: [0.067235, 0.065647, 0.067795, 0.065381, 0.06589, 0.065291, 0.068491, 0.067432, 0.069164, 0.064867]
- mean: 0.066719s, median: 0.066562s
- counters: {'total': 1, 'autograd_cache_miss': 1, 'ok': 1}, {'fxgraph_cache_miss': 1}

After:
- times: [0.072664, 0.07297, 0.072188, 0.072499, 0.073368, 0.072725, 0.072112, 0.073108, 0.072973, 0.072289]
- mean: 0.072690s, median: 0.072695s
- counters: {'total': 1, 'autograd_cache_miss': 1, 'ok': 1, 'autograd_cache_saved': 1}, {'fxgraph_cache_miss': 2}

The cache-miss path pays about +0.00613s median compile+backward time for this tiny call-free backward so it can save the AOTAutograd entry; the second run then hits AOTAutograd cache.

Test Plan:
- Reproducer before fix: first run autograd_cache_miss=1; after clearing Dynamo/codecache, second run autograd_cache_miss=2 with no autograd_cache_saved.
- Reproducer after fix: first run autograd_cache_saved=1; after clearing Dynamo/codecache, second run autograd_cache_hit=1.
- CPU reproducer after fix: same save/hit behavior and matching outputs/gradients.
- python test/dynamo/test_aot_autograd_cache.py -k lazy_backward_no_ops
- Direct compiled_fx_graph_hash check with aot_autograd_needs_fx_graph_cache_key=False vs True: same key and flag absent from debug lines.
- git diff --check && python -m py_compile torch/_inductor/codecache.py torch/_inductor/compile_fx.py torch/_functorch/_aot_autograd/autograd_cache.py test/dynamo/test_aot_autograd_cache.py
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says CI failed because direct `_compile_fx_inner()` callers, including the AOTInductor no-op logging tests, did not pass the new `aot_autograd_needs_fx_graph_cache_key` kwarg. I updated the no-op fast-path check to default that flag to false for direct callers, kept the AOTAutograd cache path behavior unchanged, and resubmitted the stack.

**Comment by jansel** [comment]
My agent says the remaining new CI failure was `test_triton_kernel_no_clones_grad_True_dynamic_True` assuming `run_and_get_code()` captured exactly one generated source. This PR now intentionally captures an extra call-free backward source when AOTAutograd is saving a cache entry, so I updated the test to inspect all captured sources for clone/copy checks and select the forward Triton source for the output-shape assertions. The targeted test now passes locally with and without `TORCHINDUCTOR_CPP_WRAPPER=1`, and the lazy-backward AOTAutograd cache tests plus `lintrunner -a` pass.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and resolved the `FxGraphHashDetails.EXCLUDED_KWARGS` conflict by preserving main’s newer hash-analysis logic while retaining this PR’s control-only kwarg exclusion. The lazy-backward cache tests, affected Triton test in normal and C++ wrapper modes, CPU miss/save/hit reproducer, syntax checks, and `lintrunner -a` all pass.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
